PyInstaller 4.5.1 "codesign failure!" on macos

I'm trying to build my application on macOS and I'm getting this error with the latest PyInstaller 4.5.1, it is working on PyInstaller 4.3

```
[2021-08-26T07:28:02.721Z] 12064 WARNING: codesign command (['codesign', '-s', '-', '--force', '--all-architectures', '--timestamp', '/Users/safebreach/Library/Application Support/pyinstaller/bincache00_py37_64bit/x86_64/adhoc/no-entitlements/libintl.8.dylib']) failed with error code 1!
[2021-08-26T07:28:02.721Z] stdout: ''
[2021-08-26T07:28:02.721Z] stderr: '/Users/safebreach/Library/Application Support/pyinstaller/bincache00_py37_64bit/x86_64/adhoc/no-entitlements/libintl.8.dylib: invalid or unsupported format for signature\n'
[2021-08-26T07:28:02.721Z] Traceback (most recent call last):
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/bin/pyinstaller", line 8, in <module>
[2021-08-26T07:28:02.721Z]     sys.exit(run())
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/__main__.py", line 126, in run
[2021-08-26T07:28:02.721Z]     run_build(pyi_config, spec_file, **vars(args))
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/__main__.py", line 65, in run_build
[2021-08-26T07:28:02.721Z]     PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/building/build_main.py", line 815, in main
[2021-08-26T07:28:02.721Z]     build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/building/build_main.py", line 762, in build
[2021-08-26T07:28:02.721Z]     exec(code, spec_namespace)
[2021-08-26T07:28:02.721Z]   File "cicd/connectivity_check.spec", line 55, in <module>
[2021-08-26T07:28:02.721Z]     console=True
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/building/api.py", line 515, in __init__
[2021-08-26T07:28:02.721Z]     entitlements_file=self.entitlements_file
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/building/api.py", line 208, in __init__
[2021-08-26T07:28:02.721Z]     self.__postinit__()
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/building/datastruct.py", line 159, in __postinit__
[2021-08-26T07:28:02.721Z]     self.assemble()
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/building/api.py", line 280, in assemble
[2021-08-26T07:28:02.721Z]     entitlements_file=self.entitlements_file)
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/building/utils.py", line 391, in checkCache
[2021-08-26T07:28:02.721Z]     osxutils.sign_binary(cachedfile, codesign_identity, entitlements_file)
[2021-08-26T07:28:02.721Z]   File "/Volumes/workspace/workspace/ConnectivityCheck_develop/connectivity_check_venv/lib/python3.7/site-packages/PyInstaller/utils/osx.py", line 383, in sign_binary
[2021-08-26T07:28:02.721Z]     raise SystemError("codesign failure!")
```

Is there a way to work around it?

PyInstaller: 4.5.1
Python: 3.7.9
OS: macOS

@rokm After commenting out the line from my env this is what I'm getting:
```
35075 INFO: Fixing EXE for code signing /Volumes/workspace/workspace/UpdateClient_master/build/update-client/updater
Traceback (most recent call last):
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/bin/pyinstaller", line 8, in <module>
    sys.exit(run())
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/__main__.py", line 126, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/__main__.py", line 65, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/building/build_main.py", line 815, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/building/build_main.py", line 762, in build
    exec(code, spec_namespace)
  File "cicd/update-client.spec", line 66, in <module>
    icon='../icon.ico'
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/building/api.py", line 524, in __init__
    self.__postinit__()
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/building/datastruct.py", line 159, in __postinit__
    self.assemble()
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/building/api.py", line 761, in assemble
    osxutils.fix_exe_for_code_signing(self.name)
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/utils/osx.py", line 212, in fix_exe_for_code_signing
    symtab_sec.stroff + symtab_sec.strsize, "Sanity check failed!"
AssertionError: Sanity check failed!
```

I'm using onedir and not onefile

This is the first command output:
```
(updater_venv) Build-Simulator-2:UpdateClient_master safebreach$ python
Python 3.7.9 (default, Sep  8 2020, 10:54:43)
[Clang 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from PyInstaller.utils.osx import fix_exe_for_code_signing, binary_to_target_arch
>>> binary_to_target_arch('./run', 'x86_64')
>>> fix_exe_for_code_signing('./run')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/utils/osx.py", line 195, in fix_exe_for_code_signing
    assert len(sign_sec) == 0, "Executable contains code signature!"
AssertionError: Executable contains code signature!
>>>
```

And the second one:
```
(updater_venv) Build-Simulator-2:UpdateClient_master safebreach$ python
Python 3.7.9 (default, Sep  8 2020, 10:54:43)
[Clang 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from PyInstaller.utils.osx import fix_exe_for_code_signing, binary_to_target_arch, remove_signature_from_binary
>>> binary_to_target_arch('./run', 'x86_64')
>>> remove_signature_from_binary('./run')
>>> fix_exe_for_code_signing('./run')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Volumes/workspace/workspace/UpdateClient_master/updater_venv/lib/python3.7/site-packages/PyInstaller/utils/osx.py", line 212, in fix_exe_for_code_signing
    symtab_sec.stroff + symtab_sec.strsize, "Sanity check failed!"
AssertionError: Sanity check failed!
>>>
```

--- a/PyInstaller/building/utils.py
+++ b/PyInstaller/building/utils.py
@@ -369,11 +369,13 @@ def checkCache(
     cache_index[basenm] = digest
     misc.save_py_data_struct(cacheindexfn, cache_index)
 
-    # On Mac OS we need relative paths to dll dependencies starting with @executable_path. Modifying headers
-    # invalidates signatures, so remove any existing signature and then re-add it after paths are rewritten.
+    # On Mac OS we need relative paths to dll dependencies starting with @executable_path. While modifying
+    # the headers invalidates existing signatures, we avoid removing them in order to speed things up (and
+    # to avoid potential bugs in the codesign utility, like the one reported on Mac OS 10.13 in #6167).
+    # The forced re-signing at the end should take care of the invalidated signatures.
     if is_darwin:
         osxutils.binary_to_target_arch(cachedfile, target_arch, display_name=fnm)
-        osxutils.remove_signature_from_binary(cachedfile)
+        #osxutils.remove_signature_from_binary(cachedfile)  # Disabled as per comment above.
         dylib.mac_set_relative_dylib_deps(cachedfile, dist_nm)
         osxutils.sign_binary(cachedfile, codesign_identity, entitlements_file)
     return cachedfile

--- a/PyInstaller/utils/osx.py
+++ b/PyInstaller/utils/osx.py
@@ -186,18 +186,22 @@ def fix_exe_for_code_signing(filename):
     symtab_sec = [cmd for cmd in header.commands if cmd[0].cmd == LC_SYMTAB]
     assert len(symtab_sec) == 1, "Expected exactly one SYMTAB section!"
     symtab_sec = symtab_sec[0][1]  # Take the symtab command entry
-    # Sanity check; the string table is located at the end of the SYMTAB section, which in turn is the last section in
-    # the __LINKEDIT segment.
-    assert linkedit_seg.fileoff + linkedit_seg.filesize == symtab_sec.stroff + symtab_sec.strsize, \
-        "Sanity check failed!"
 
-    # Compute the old/declared file size (header.offset is zero for single-arch thin binaries).
-    old_file_size = header.offset + linkedit_seg.fileoff + linkedit_seg.filesize
-    delta = file_size - old_file_size
-    # Expand the string table in SYMTAB section...
-    symtab_sec.strsize += delta
-    # .. as well as its parent __LINEDIT segment.
-    linkedit_seg.filesize += delta
+    # The string table is located at the end of the SYMTAB section, which in turn is the last section in the __LINKEDIT
+    # segment. Therefore, the end of SYMTAB section should be aligned with the end of __LINKEDIT segment, and in turn
+    # both should be aligned with the end of the file (as we are in the last or the only arch slice).
+    #
+    # However, when removing the signature from the executable using codesign under Mac OS 10.13, the codesign utility
+    # may produce an invalid file, with declared length of __LINKEDIT segment (linkedit_seg.filesize), as reported in
+    # issue #6167.
+    #
+    # We can compensate for that by not using the declared sizes anywhere, and simply recompute them. In the final
+    # binary, the __LINKEDIT segment and the SYMTAB section MUST end at the end of the file (otherwise, we have bigger
+    # issues...). So simply recompute the declared sizes as difference between the final file length and the
+    # corresponding start offset.
+    symtab_sec.strsize = file_size - symtab_sec.stroff
+    linkedit_seg.filesize = file_size - linkedit_seg.fileoff
+
     # Compute new vmsize by rounding filesize up to full page size.
     page_size = (0x4000 if _get_arch_string(header.header).startswith('arm64') else 0x1000)
     linkedit_seg.vmsize = math.ceil(linkedit_seg.filesize / page_size) * page_size

--- a/PyInstaller/utils/osx.py
+++ b/PyInstaller/utils/osx.py
@@ -192,8 +192,8 @@ def fix_exe_for_code_signing(filename):
     # both should be aligned with the end of the file (as we are in the last or the only arch slice).
     #
     # However, when removing the signature from the executable using codesign under Mac OS 10.13, the codesign utility
-    # may produce an invalid file, with declared length of __LINKEDIT segment (linkedit_seg.filesize), as reported in
-    # issue #6167.
+    # may produce an invalid file, with the declared length of the __LINKEDIT segment (linkedit_seg.filesize) pointing
+    # beyond the end of file, as reported in issue #6167.
     #
     # We can compensate for that by not using the declared sizes anywhere, and simply recompute them. In the final
     # binary, the __LINKEDIT segment and the SYMTAB section MUST end at the end of the file (otherwise, we have bigger

--- a/news/6167.bugfix.rst
+++ b/news/6167.bugfix.rst
@@ -1,0 +1,2 @@
+(macOS) Robustify the macOS assembly pipeline to work around the issues with 
+the ``codesign`` utility on macOS 10.13 High Sierra.

